libkrun/libkruntime.c

```diff
@@ -342,126 +419,213 @@ libkruntime_done(void)
+ * TSR value to cause precision loss (you would need a integer that would not
+ * fit in a 52-bit unsigned int before precision starts being lost in
+ * lower-order bits). Nevertheless, we check it.
+ *
+ * This routine comes at a small cost (a handful of asm instrs). Note that this
+ * cost is a drop in the ocean compared to benchmark workloads.
  */
-uint64_t
-sum_msr_all_cores(long msr_addr, uint64_t mask)
+double
+krun_u64_to_double(uint64_t u64_val)
+{
+    double d_val = (double) u64_val;
+    uint64_t u64_val2 = (uint64_t) d_val;
+
+    if (u64_val != u64_val2) {
+        fprintf(stderr, "Loss of precision detected!\n");
+        fprintf(stderr, "%" PRIu64 " != %" PRIu64 "\n", u64_val, u64_val2);
+        exit(EXIT_FAILURE);
+    }
+    // (otherwise all is well)
+    return d_val;
+}
+
+/*
+ * Load up the results struct.
+ */
+void
+krun_measure(int mdata_idx)
 {
-    int core;
-    uint64_t sum = 0, val;
+    struct krun_data *data = &(krun_mdata[mdata_idx]);
 
-    for (core = 0; core < num_msr_nodes; core++) {
-        val = read_msr(core, msr_addr) & mask;
+    KDATA_BOUNDS_CHECK(mdata_idx);
 
-        // Check for overflow
-        if (sum > UINT64_MAX - val) {
-            fprintf(stderr, "Sum overflow!: "
-                    "%" PRIu64 " + %" PRIu64 "\n", sum, val);
-            exit(EXIT_FAILURE);
+#if defined(__linux__) && !defined(TRAVIS)
+    // We support per-core readings
+
+    // We do it this way to ensure correct nesting of reads
+    if (mdata_idx == 0) {
+        // start readings
+        for (int core = 0; core < krun_num_cores; core++) {
+            data->aperf[core] = krun_read_aperf(core);
+            data->mperf[core] = krun_read_mperf(core);
+            data->core_cycles[core] = krun_read_core_cycles(core);
+        }
+        data->wallclock = krun_clock_gettime_monotonic();
+    } else {
+        // stop readings
+        data->wallclock = krun_clock_gettime_monotonic();
+        for (int core = 0; core < krun_num_cores; core++) {
+            data->core_cycles[core] = krun_read_core_cycles(core);
+            data->aperf[core] = krun_read_aperf(core);
+            data->mperf[core] = krun_read_mperf(core);
         }
-        sum += val;
+        krun_check_mdata();
     }
-
-    return sum;
-}
 #elif defined(__linux__) && defined(TRAVIS)
-// function not used on non-linux or travis
+    data->wallclock = krun_clock_gettime_monotonic();
+    if (mdata_idx == 1) {
+        krun_check_mdata();
+    }
+#elif defined(__OpenBSD__)
+    data->wallclock = krun_clock_gettime_monotonic();
+    if (mdata_idx == 1) {
+        krun_check_mdata();
+    }
 #else
 #error "Unsupported platform"
-#endif // __linux__ && && !TRAVIS
+#endif
+}
 
 /*
- * For languages like Lua, where there is suitible integer type
+ * Check all of the measurements for issues.
  */
-double
-read_core_cycles_double()
+void
+krun_check_mdata(void)
 {
-    return u64_to_double(read_core_cycles());
+    int bail = 0, core;
```

Comment:
`bail` seems pointless: just call exit (or do a goto) at the appropriate point, rather than printing out lots of errors?